pkg/controller/elasticsearch/pdb/reconcile_with_roles.go

```diff
@@ -0,0 +1,560 @@
+		}
+		for _, idx := range indices {
+			if visited[idx] {
+				continue
+			}
+			targetPDBRole := role
+			// if we already assigned a PDB for this role, use that instead
+			if target, ok := roleToTargetPDB[role]; ok {
+				targetPDBRole = target
+			}
+			grouped[targetPDBRole] = append(grouped[targetPDBRole], idx)
+			for _, r := range indicesToRoles[idx].AsSlice() {
+				roleToTargetPDB[esv1.NodeRole(r)] = targetPDBRole
+			}
+			visited[idx] = true
+		}
+	}
+	// transform into the expected format
+	res := make(map[esv1.NodeRole][]appsv1.StatefulSet)
+	for role, indices := range grouped {
+		group := make([]appsv1.StatefulSet, 0, len(indices))
+		for _, idx := range indices {
+			group = append(group, statefulSets[idx])
+		}
+		res[role] = group
+	}
+	return res, nil
+}
+
+// getPrimaryRoleForPDB returns the primary role from a set of roles for PDB naming and grouping.
+// Data roles are most restrictive (require green health), so they take priority.
+// All other roles have similar disruption rules (require yellow+ health).
+func getPrimaryRoleForPDB(roles sets.Set[esv1.NodeRole]) esv1.NodeRole {
+	if len(roles) == 0 {
+		return "" // coordinating role
+	}
+
+	// Data roles are most restrictive (require green health), so they take priority.
+	// Check if any data role variant is present (excluding data_frozen)
+	if slices.ContainsFunc(dataRoles, func(dataRole esv1.NodeRole) bool {
+		return roles.Has(dataRole)
+	}) {
+		// Return generic data role for all data role variants
+		return esv1.DataRole
+	}
+
+	// Master role comes next in priority
+	if _, ok := roles[esv1.MasterRole]; ok {
+		return esv1.MasterRole
+	}
+
+	// Data frozen role (has different disruption rules than other data roles)
+	if _, ok := roles[esv1.DataFrozenRole]; ok {
+		return esv1.DataFrozenRole
+	}
+
+	// Return the first role we encounter in a deterministic order
+	// Define a priority order for non-data roles
+	nonDataRoles := []esv1.NodeRole{
+		esv1.IngestRole,
+		esv1.MLRole,
+		esv1.TransformRole,
+		esv1.RemoteClusterClientRole,
+	}
+
+	// Check non-data roles in priority order
+	for _, role := range nonDataRoles {
+		if _, ok := roles[role]; ok {
+			return role
+		}
+	}
+
+	// If no known role found, return any role from the map
+	for role := range roles {
+		return role
+	}
+
+	// Should never reach here if roles is not empty
+	return ""
+}
+
+// getRolesForStatefulSet gets the roles from a StatefulSet's expected configuration.
+func getRolesForStatefulSet(
+	statefulSet appsv1.StatefulSet,
+	expectedResources nodespec.ResourcesList,
+	v version.Version,
+) ([]esv1.NodeRole, error) {
+	forStatefulSet, err := expectedResources.ForStatefulSet(statefulSet.Name)
+	if err != nil {
+		return nil, err
+	}
+	cfg, err := forStatefulSet.Config.Unpack(v)
+	if err != nil {
+		return nil, err
+	}
+	nodeRoles := make([]esv1.NodeRole, len(cfg.Node.Roles))
+	for i, role := range cfg.Node.Roles {
+		nodeRoles[i] = esv1.NodeRole(role)
+	}
+	return nodeRoles, nil
```

Comment:
I don't think this is correct. By default not specifying any roles means the node has all roles. We have logic implemented already that takes that into account when calculating the labels for the statefulset. 
With this logic here we would treat a nodeSet without explicit roles incorrectly as coordinating. However the only way to create a coordinating only nodeSet is by setting `node.roles: []`

Reply:
Good catch here. This is a wrinkle I wasn't aware of, tbh. I've fixed the issue, and added additional unit tests to ensure this is handled properly.